// oas/dist/extensions.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});














var _chunkXS7VDTTTcjs = require('./chunk-XS7VDTTT.cjs');















exports.CODE_SAMPLES = _chunkXS7VDTTTcjs.CODE_SAMPLES; exports.DISABLE_TAG_SORTING = _chunkXS7VDTTTcjs.DISABLE_TAG_SORTING; exports.EXPLORER_ENABLED = _chunkXS7VDTTTcjs.EXPLORER_ENABLED; exports.HEADERS = _chunkXS7VDTTTcjs.HEADERS; exports.METRICS_ENABLED = _chunkXS7VDTTTcjs.METRICS_ENABLED; exports.OAUTH_OPTIONS = _chunkXS7VDTTTcjs.OAUTH_OPTIONS; exports.PARAMETER_ORDERING = _chunkXS7VDTTTcjs.PARAMETER_ORDERING; exports.PROXY_ENABLED = _chunkXS7VDTTTcjs.PROXY_ENABLED; exports.SAMPLES_LANGUAGES = _chunkXS7VDTTTcjs.SAMPLES_LANGUAGES; exports.SIMPLE_MODE = _chunkXS7VDTTTcjs.SIMPLE_MODE; exports.extensionDefaults = _chunkXS7VDTTTcjs.extensionDefaults; exports.getExtension = _chunkXS7VDTTTcjs.getExtension; exports.hasRootExtension = _chunkXS7VDTTTcjs.hasRootExtension; exports.validateParameterOrdering = _chunkXS7VDTTTcjs.validateParameterOrdering;
